source/MicroBitMemoryMap.cpp

```diff
@@ -138,5 +140,58 @@ void MicroBitMemoryMap::findHashes()
                 memoryMapStore.memoryMap[2].startAddress = (uint32_t)magicAddress;
                 return;
         }
+        
+        if (   *(magicAddress - (sizeof(magicAddress))) == 0x597F30FE
+            && *(magicAddress - (sizeof(magicAddress)/4))   == 0xC1B1D79D
+           ) {
+            // Found uPy Magic
+            DMESG("FOUND UPY_MAGIC");
+            uint8_t nRegions = (*(magicAddress - (sizeof(magicAddress)/2)));
+            
+            // Iterate through regions
+            for(int i = 0; i < nRegions; i++) {
+               processRecord(magicAddress - ((2 + i) * sizeof(magicAddress))); 
+            }
+
+        }
```

Comment:
With all this pointer arithmetic maybe it's getting to the point where it might be more clear if there was a struct defined with the different entries being members of the expected size.
We can probably do the same with the record rows.

Reply:
Yeah - that's a good idea!